docs/specs/parsing.md

```diff
@@ -13,18 +18,75 @@
     - `DateTime`, `DateTimeOffset`
     - `TimeSpan`
     - `Guid`
+    - `Uri`
 * Parses different cultures
     - e.g. English and German dates
-    - Needs to handle all cultures .NET can handle
+    - Needs to handle all cultures .NET Core can handle
+* Parse any outputs produced by [formatting](formatting.md)
+    - This ensures parsing & formatting behave symmetric
 * Parse from these encodings
-    - UTF8      <------------- THATS WHAT WE CARE ABOUT
+    - UTF8
     - ASCII
     - UTF16-BE
     - UTF16-LE
 * Parses from existing buffers without knowing up-front how many bytes the
   parsing needs to consume
 
-## Discussion
+## Focus on UTF8
+
+While we generally need to handle parsing for more than just UTF8 we belive it's
+beneficial to have an implemetnation that is and-optimized for UTF8 instead of
+having an implementation that is going through some abstraction.
+
+That's because UTF8 is the dominant encoding for cloud-based apps. We'd support
+other encodings by either converting them to the UTF8 before or by having an
+implementation that can parse different encodings through an abstraction.
+
+## Low-Level API Shape
+
+We generally want the APIs to be static methods and being able to operate on
+raw data, which could either be native memory or a managed (byte) array.
+
+Using the [Span\<T>.md] feature we can express this quite easily using
+`ReadOnlySpan<T>` (`Xyz` indicates the name of the type being parsed):
+
+```C#
+public static bool TryParseXyz(ReadOnlySpan<byte> text,
+                               out Xyz value,
+                               out int bytesConsumed)
+```
+
+This allows the parsing to happen from a wide variety of data sources:
+
+* **Native memory**, by creating a span from the pointer + lengh.
+* **`byte[]`**, via an implicit conversion
+* **`Span<byte>`**, via an implicit conversion
+
+Depending on how we express encodings and cultures, it's likeky what we need to
+create additional arguments, defaulted to being UTF8 and invariant.
+
+The result of the method indicates whether parsing was successful. Since the
+entire point of these APIs is performance, we don't plan on adding APIs that
+throw exceptions when parsing cannot succeed.
+
+## High-Level API Shape
+
+The problem with the low-level API shape is that it doesn't deal with situations
+where the data spans multiple buffers. Thus, we need a way to for the parsing
+methods to "pull" the next buffer. We currently have something like
+`ISequence<T>`, which is basically a more efficient `IEnumerable<T>`:
+
+```C#
+public static bool TryParseXyz<T>(this T memorySequence,
+                                  out Xyz value,
+                                  out int bytesConsumed)
+    where T : ISequence<ReadOnlyMemory<byte>>;
+```
+
+We'll need to decide if that's the right approach or not. Alternatively, we
```

Comment:
The solution without language/runtime enforced safety seems to have a strong managed/unmanaged interop bias. So maybe a reasonable compromise is:

- Move Memory/OwnerMemory to System.Runtime.InteropServices namespace. This namespace is full of similar types that you need to be careful about using correctly.
- Make it non-generic, for byte buffers only. Generics do not make a lot of sense for maanaged/unmanaged interop.
- Consider reusing the existing `System.Runtime.InteropServices.SafeBuffer` for OwnedMemory because they are essentially the same types.